data_loader.py

```diff
@@ -28,45 +31,103 @@ def disconnect(self):
+            for k in data:
+                yield from self._flatten_dict(data[k], keystring + str(k))
+        else:
+            yield keystring, data
+
+    def _combine_excel(self, list_of_articles: list):
+        """Add excel values to all articles"""
+        review_status = pd.read_excel(
+            self.args.reference_data_file,
+            sheet_name="review_status",
+            header=2,
+            usecols="B:E",
+            dtype=str,
+        )
+        date_completed = pd.read_excel(
+            self.args.reference_data_file, sheet_name="date_completed"
+        )
+
+        # Preprocessing data
+        review_status.columns = review_status.columns.str.replace(" ", "_").str.lower()
+        date_completed.columns = date_completed.columns.str.replace(
+            " ", "_"
+        ).str.lower()
+        review_status["article_id"] = review_status["article_id"].apply(
+            lambda x: x.strip()
+        )
+        date_completed["reference_id"] = date_completed["reference_id"].values.astype(
+            str
+        )
+
+        for article in list_of_articles:
+            excel_dict = {}
+            article_status = review_status.loc[
+                review_status["article_id"] == article["_id"]
+            ]
+            if len(article_status) > 1:
+                excel_dict["status.duplicates"] = len(article_status)
+                excel_dict["status.duplicates.reference_id"] = article_status[
+                    "reference_id"
+                ].tolist()
+
+            if article_status.empty:
+                excel_dict.update(article_status.to_dict())
+                reference_id = None
+            else:
+                excel_dict.update(article_status.tail(1).to_dict("records")[0])
+                reference_id = article_status["reference_id"].values
+
+            article_date = date_completed.loc[
+                date_completed["reference_id"].isin(reference_id)
+            ]
+
+            if article_date.empty:
+                excel_dict.update(article_date.to_dict())
+            else:
+                excel_dict.update(article_date.tail(1).to_dict("records")[0])
+
+            article.update(excel_dict)
+
+        return list_of_articles
+
     def getDataBatch(self, batch_size):
         """
         Generator - Get data from source on batches.
 
         :returns One list for each batch. Each of those is a list of
                  dictionaries with the defined rows.
         """
-        # TODO: implement - this dummy implementation returns one batch of data
-        yield [
-            {
-                "headline.main": "The main headline",
-                "_id": "1234",
-            }
+        with open(self.args.api_response_file) as json_file:
+            data = json.load(json_file)
+
+        list_of_articles = [
+            {k: v for k, v in self._flatten_dict(article)}
+            for article in self._combine_excel(data["response"]["docs"])
         ]
```

Comment:
Not sure if it's just me but nested comprehensions are hard to read ;)

Reply:
Also, it first does the operations like combining excel and flatten dictionaries for all data and only when all the data are processed it starts yielding batches.
The idea of this function being a generator is to stream data continuously and not waiting for all the data being processed.
Let's say the response returns 1milion data which has to be processed. Then we need to wait for all data and only then we receive the first batch.
I would see it as this:
```python
batch = []
for article in self._get_articles():
    combined_article = self._combine_excel(article)
    flatten = self._flatten_dict(combined_article)
    batch.append(flatten)
    if len(batch) >= batch_size:
        yield batch
        batch = []
if batch:
   yield batch
````

Reply:
And I would do `flatten` before `combine_excel` because inside `combine_excel` you create new keys using dot which could mean that dictionary is already flatten.